### DISCUSSION TITLE
How to generate geotif labels from a image with a trained model?

### DISCUSSION THREAD

**Discussion by lcoandrade** [torchgeo/torchgeo#1470] [Q&A] [answered]
Hi there!

I'm working on a Kaggle [notebook](https://www.kaggle.com/code/luizclaudioandrade/torchgeo-101) with a custom [dataset](https://www.kaggle.com/datasets/luizclaudioandrade/naip-chesapeake-sample) based on NAIP images and Chesapeake labels.

I want to perform a segmentation task.

Right now, I'm able to train with one GPU. How can I, from this trained model, generate geotif labels from a image?

**Comment by adamjstewart** [comment]
@adriantre is working on a way to automatically do this in SemanticSegmentationTask in #1407. If you just want a PNG, it's really easy to do. If you want a GeoTIFF, we have a little bit of work required to store the CRS/transform needed to save geospatial metadata. You can also do this outside of Lightning by loading the model and data loader and iterating over inputs, making predictions, and saving to disk using the same rasterio profile.

**Reply by lcoandrade** [reply]
Is there some tutorial to follow with those Rasterio steps?

**Reply by adriantre** [reply]
[This](https://rasterio.readthedocs.io/en/stable/topics/writing.html) is where to look. Are you using GridGeoSampler, or are you passing the full image into your model? If you are passing patches into the model you need to write those output masks into the [correct spot in the original image pixel extent](https://rasterio.readthedocs.io/en/stable/topics/windowed-rw.html#writing), or save georeferenced tif files per sample with the correct window transform.

```python
bbox = ...  # `RasterDataset.__getitem__` passes query (window extent in CRS) with the key 'bbox' along with the sample
window_transform = rasterio.transform.from_bounds(*bbox[:4],  patch_size, patch_size)
```

**Reply by lcoandrade** [reply]
Thanks!

**Reply by lcoandrade** [reply]
I'm just trying to see the patches prediction for now. I'm using this code to make the predictions and to plot them:
To plot, I'm using this:
```
import torchvision.transforms as transforms
reverse_transform = transforms.Compose([
            transforms.ToPILImage(),
            transforms.Resize(size=IMG_SIZE),
        ])

# helper function for data visualization
def visualize(**images):
    """PLot images in one row."""
    n = len(images)
    plt.figure(figsize=(16, 5))
    for i, (name, image) in enumerate(images.items()):
        plt.subplot(1, n, i + 1)
        plt.xticks([])
        plt.yticks([])
        plt.title(' '.join(name.split('_')).title())
        plt.imshow(image)
    plt.show()
```

To predict I'm using this:
```
import torch

test_dataset = datamodule.test_dataset
test_sampler = datamodule.test_sampler
test_dataloader = DataLoader(test_dataset, sampler=test_sampler, collate_fn=stack_samples)

n = 0
for sample in test_dataloader:
    if n == 10:
        break
    image, gt_mask = sample['image'], sample['mask']
    
    prediction = task.model.predict(image)
    prediction = torch.softmax(prediction, dim=1)
    prediction = torch.argmax(prediction.squeeze(), dim = 0)
    
    visualize(
        image=reverse_transform(image.squeeze()),
        ground_truth = gt_mask.squeeze(),
        prediction = prediction,
    )
    n += 1
```

But the results seem to be way different from what should be expected for values like these:
```
[{'test_loss': 6.3207671701093204e-06,
  'test_MulticlassAccuracy': 0.9367654323577881,
  'test_MulticlassJaccardIndex': 0.8810524940490723}]
```

This is what a get as predictions:
<img width="1066" alt="Captura de Tela 2023-07-12 às 14 18 23" src="https://github.com/microsoft/torchgeo/assets/8769816/f27290f2-e1a0-48b6-83ab-437a70b87392">

What am I doing wrong?

**Reply by calebrob6** [reply]
Hey @lcoandrade,

A few things to check:
- It looks like the image is shown in BGR format with some transparency. Can you try visualizing `image[:,:,[2,1,0]]`?
- If you do not set a max/min, or a set colormap on `plt.imshow(...)` then it will show different colors for different classes. Try something like: (edit to set the vmin and vmax value appropriately)
```
def visualize(**images):
    """Plot images in one row."""
    fig, axs = plt.subplots(1, 3, figsize=(16, 5))
    axs[0].axis("off")
    axs[1].axis("off")
    axs[2].axis("off")
    
    axs[0].imshow(images["image"][:,:,[2,1,0]])
    axs[1].imshow(images["ground_truth"], vmin=0, vmax=4, interpolation="none")
    axs[2].imshow(images["prediction"], vmin=0, vmax=4, interpolation="none")
    
    plt.show()
    plt.close()
```

**Reply by lcoandrade** [reply]
I've made some changes:
1 - I'm using a colormap.
2 - I'm using a random sampler on my test dataset to change the patches sent to the model.
3 - I'm plotting the images changing the band order to [2,1,0] to avoid the transparency.

I've created a colormap to apply on the masks and predictions. This is is function to apply the colormap:
```
colormap = [
    [0, 255, 255], [255, 255, 0], [255, 0, 255], [0, 255, 0], [0, 0, 255], [255, 255, 255], [0, 0, 0],
    [0, 100, 100], [100, 100, 0], [100, 0, 100], [0, 100, 0], [0, 0, 100], [100, 100, 100]
           ]

# Perform colour coding on the outputs
def colour_code_segmentation(image, colormap):
    """
    Given a 1-channel array of class keys, colour code the segmentation results.
    # Arguments
        image: single channel array where each value represents the class key.
        colormap: the list os rgb colors for each class

    # Returns
        Colour coded image for segmentation visualization
    """
    colour_codes = np.array(colormap)
    return colour_codes[image]
```

To plot the images, I'm using this updated code:
```
import torch
import numpy as np

test_dataset = datamodule.test_dataset
test_sampler = RandomGeoSampler(test_dataset, size=IMG_SIZE, length=SAMPLE_SIZE)
test_dataloader = DataLoader(test_dataset, sampler=test_sampler, collate_fn=stack_samples)

n = 0
for sample in test_dataloader:
    if n == 10:
        break
    image, gt_mask = sample['image'], sample['mask']
    
    prediction = task.model.predict(image)
    prediction = torch.softmax(prediction, dim=1)
    prediction = torch.argmax(prediction, dim = 1)
    
    # reapplying the original colors in the reversed one hot images
    gt_mask = colour_code_segmentation(gt_mask, colormap)
    prediction = colour_code_segmentation(prediction, colormap)
        
    visualize(
        image=reverse_transform(image.squeeze()[[2,1,0]]),
        ground_truth = gt_mask.squeeze(),
        prediction = prediction.squeeze(),
    )
    n += 1
```

Here is the updated [notebook](https://www.kaggle.com/code/luizclaudioandrade/torchgeo-101)

But even with the color map, the predictions are coming out quite strange.

I'm lost here...

Could you check the notebook to see where should I change the code to make the predictions come out properly?

**Reply by calebrob6** [reply]
I don't see the updated code in the notebook.

**Reply by lcoandrade** [reply]
I'm still committing.... 🤣
I wasn't expecting such a quick answer. In a few more minutes the version 9 will be ready.

Thanks!

**Reply by calebrob6** [reply]
Ah I see :) -- you just happened to be at the very top of my inbox, I'll check back in 30 minutes

**Reply by lcoandrade** [reply]
Version 9 is ready for review now.
Thanks!

**Reply by calebrob6** [reply]
In cell [16] I would try dividing image by 255.0

In GeoDataModule augmentation is done when batches are transfered to the GPU, but in cell [16] you are bypassing this by creating a dataloader from the datamodule's test dataset so I don't think augmentation is being applied.

Also, something is still wrong with the images... maybe just try `image[:,:,:3]` instead of `image[:,:,[2,1,0]]`?

**Reply by calebrob6** [reply]
More generally, if your model output is somewhat constant, or definitely not what you expect, a good first debugging step is to examine the inputs closely and see if they are aligning with the type of inputs that the model was trained on. E.g. if the model was trained on inputs where you have done standardization per-channel, it is important to do this same step at inference time as well.

**Reply by lcoandrade** [reply]
I understand, but I'm treating the images used to train and those used in the prediction the same way. I just create a dataset with the NAIP images and Chesapeake labels and make a random split to generate train, validation and test datasets.

Custom datasets:
```
class NAIPImages(RasterDataset):
    filename_glob = "m_*.tif"
    is_image = True
    separate_files = False
    
class ChesapeakeLabels(RasterDataset):
    filename_glob = "m_*.tif"
    is_image = False
    separate_files = False
```

Making them individually:
```
naip_root = os.path.join(INPUT_DIR, 'naip_images')
naip_images = NAIPImages(naip_root)
print(naip_images)

chesapeake_root = os.path.join(INPUT_DIR, "chesapeake_labels")
chesapeake_labels = ChesapeakeLabels(chesapeake_root)
print(chesapeake_labels)
```

Creating an intersection:
```
dataset = naip_images & chesapeake_labels # this means I'm creating an IntersectionDataset
```

From this, I create my custom datamodule that is used to train my network:
```
class CustomGeoDataModule(GeoDataModule):
    def setup(self, stage: str) -> None:
        """Set up datasets.

        Args:
            stage: Either 'fit', 'validate', 'test', or 'predict'.
        """
        self.dataset = self.dataset_class(**self.kwargs)
        
        generator = torch.Generator().manual_seed(0)
        (
            self.train_dataset,
            self.val_dataset,
            self.test_dataset,
        ) = random_bbox_assignment(dataset, [0.6, 0.2, 0.2], generator)
        
        if stage in ["fit"]:
            self.train_batch_sampler = RandomBatchGeoSampler(
                self.train_dataset, self.patch_size, self.batch_size, self.length
            )
        if stage in ["fit", "validate"]:
            self.val_sampler = GridGeoSampler(
                self.val_dataset, self.patch_size, self.patch_size
            )
        if stage in ["test"]:
            self.test_sampler = GridGeoSampler(
                self.test_dataset, self.patch_size, self.patch_size
            )
            
datamodule = CustomGeoDataModule(
    dataset_class = type(dataset), # GeoDataModule kwargs
    batch_size = 10, # GeoDataModule kwargs
    patch_size = IMG_SIZE, # GeoDataModule kwargs
    length = SAMPLE_SIZE, # GeoDataModule kwargs
    num_workers = WORKERS, # GeoDataModule kwargs
    dataset1 = naip_images, # IntersectionDataset kwargs
    dataset2 = chesapeake_labels, # IntersectionDataset kwargs
    collate_fn = stack_samples, # IntersectionDataset kwargs
)
```

This datamodule is used to train and to predict.

**Reply by calebrob6** [reply, marked as answer]
I know, but when you do:

```
trainer.test(
    model=task,
    datamodule=datamodule
)
```

I'm pretttty sure the images are being divided by 255

and when you do:
```
test_dataset = datamodule.test_dataset
test_sampler = RandomGeoSampler(test_dataset, size=IMG_SIZE, length=SAMPLE_SIZE)
test_dataloader = DataLoader(test_dataset, sampler=test_sampler, collate_fn=stack_samples)

n = 0
for sample in test_dataloader:
    ...
```

I'm pretty sure the images are _not_ being divided by 255.

**Reply by lcoandrade** [reply]
Ahhhh, I'll insert an:
```
image = image/255.
```

To see what will happen.

**Reply by calebrob6** [reply]
Also change up how the images are being visualized ;)

**Reply by lcoandrade** [reply]
Just making:
```
image = image/255.
```

Solved the problem!

Thank you very much!

Now I would like to proceed to generate a geotiff for a whole image.

**Reply by lcoandrade** [reply]
Version 10 is being commit to show the correct outputs!

**Reply by calebrob6** [reply]
Great 😄 -- please check out Adam and Adrian's messages above and feel free to message more if you have further questions

**Reply by lcoandrade** [reply]
The images were solved with:
```
image.squeeze()[:3]
```

**Reply by adriantre** [reply]
Side-note, is there a TorchGeo slack/slack-channel?

**Reply by calebrob6** [reply]
Yep, if you want to send me your email (caleb.robinson@microsoft.com) I can add you!
